lib/service/servicecfg/auth.go

```diff
@@ -189,22 +189,97 @@ type KeystoreConfig struct {
 	AWSKMS AWSKMSConfig
 }
 
+// CheckAndSetDefaults checks that required parameters of the config are
+// properly set and sets defaults.
+func (cfg *KeystoreConfig) CheckAndSetDefaults() error {
+	count := 0
+	if cfg.PKCS11 != (PKCS11Config{}) {
+		if err := cfg.PKCS11.CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err, "validiting pkcs11 config")
+		}
+		count++
+	}
+	if cfg.GCPKMS != (GCPKMSConfig{}) {
+		if err := cfg.GCPKMS.CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err, "validating gcp_kms config")
+		}
+	}
+	if cfg.AWSKMS != (AWSKMSConfig{}) {
+		if err := cfg.AWSKMS.CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err, "validating aws_kms config")
+		}
+		count++
+	}
+	if count > 1 {
+		return trace.BadParameter("must configure at most one of pkcs11, gcp_kms, or aws_kms")
+	}
+	return nil
+}
+
+// PKCS11Config holds static configuration options for a PKCS#11 HSM.
 type PKCS11Config struct {
-	Path       string
+	// Path is the path to the PKCS11 module.
+	Path string
+	// SlotNumber is the PKCS11 slot to use.
 	SlotNumber *int
+	// TokenLabel is the label of the PKCS11 token to use.
 	TokenLabel string
-	Pin        string
-	HostUUID   string
+	// Pin is the PKCS11 pin for the given token.
+	Pin string
 }
 
+// CheckAndSetDefaults checks that required parameters of the config are
+// properly set and sets defaults.
+func (cfg *PKCS11Config) CheckAndSetDefaults() error {
+	if cfg.Path == "" {
+		return trace.BadParameter("must provide Path")
+	}
+	if cfg.SlotNumber == nil && cfg.TokenLabel == "" {
+		return trace.BadParameter("must provide one of SlotNumber or TokenLabel")
+	}
+	return nil
+}
+
+// GCPKMSConfig holds static configuration options for GCP KMS.
 type GCPKMSConfig struct {
-	KeyRing         string
+	// KeyRing is the fully qualified name of the GCP KMS keyring.
+	KeyRing string
+	// ProtectionLevel specifies how cryptographic operations are performed.
+	// For more information, see https://cloud.google.com/kms/docs/algorithms#protection_levels
+	// Supported options are "HSM" and "SOFTWARE".
 	ProtectionLevel string
-	HostUUID        string
 }
 
+// CheckAndSetDefaults checks that required parameters of the config are
+// properly set and sets defaults.
+func (cfg *GCPKMSConfig) CheckAndSetDefaults() error {
+	if cfg.KeyRing == "" {
+		return trace.BadParameter("must provide a valid KeyRing")
+	}
+	switch cfg.ProtectionLevel {
+	case "SOFTWARE", "HSM":
```

Comment:
Should we make these into public constants, so they may be reused? Or is that not relevant?

Reply:
yeah probably, done